pydatastructs/linear_data_structures/tests/test_algorithms.py

```diff
@@ -126,3 +127,78 @@ def test_longest_common_sequence():
     Z = ODA(int, [])
     output = longest_common_subsequence(Y, Z)
     assert str(output) == '[]'
+
+def test_upper_bound():
+    ODA = OneDimensionalArray
+    arr1 = ODA(int, [3, 3, 3])
+    output = upper_bound(arr1, 0, len(arr1) - 1, 3, None)
+    expected_result = 2
+    assert expected_result == output
```

Comment:
From, here I think there needs a slight change in the algo. That is output `2` isn't correct mathematically for this in the given range. Hence, we should consider the interval as [start, end). And so, I expect the output of it as `3`. (You can verify it by yourself, we are making it similar to cpp)

Reply:
Also, after making this interval change, please add tests apart from [0, len(arr)) interval too.